### PR TITLE
Support quiesce, memory in vmware guest snapshot

### LINKED ISSUE #26270: vmware_guest_snapshot should allow memory dump and quiese of FS
##### ISSUE TYPE

 - Feature Idea

##### COMPONENT NAME

vmware_guest_snapshot.py

##### ANSIBLE VERSION

```
ansible 2.3.1.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
  python version = 2.7.12 (default, Nov 19 2016, 06:48:10) [GCC 5.4.0 20160609]
```

##### OS / ENVIRONMENT

Ubuntu 16.04 64 bits

##### SUMMARY

currently snapshot operations call CreateSnapshot from pyvmomi (https://github.com/vmware/pyvmomi/blob/575ab56eb56f32f53c98f40b9b496c6219c161da/docs/vim/VirtualMachine.rst) with flags for memory dump and filesystem quiece in False. This leads to every snapshot of a VM be in a powered down state irregardless of the state of the VM when the snapshot is taken.
Having these flags in False makes for a really fast snapshot procedure, but I belive it is less usefull.

##### STEPS TO REPRODUCE

dummy_vm is powered on

```yaml
- name: Create snapshot
  vmware_guest_snapshot:
    hostname: 192.168.1.209
    username: administrator@vsphere.local
    password: vmware
    name: dummy_vm
    state: present
    snapshot_name: snap1
    description: snap1_description
  delegate_to: localhost
```

##### EXPECTED RESULTS

restoring snapshot will result in dummy_vm in powered on state.

##### ACTUAL RESULTS

dummy_vm is restored in powered off state.

```

```

### PR DESCRIPTION
**Pull request by Akasurde**
##### SUMMARY
Fix adds support for quiesce and memory options while taking
snapshot of virtual machine. Update documentation and examples
for reflecting this change.

Fixes #26270

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
lib/ansible/modules/cloud/vmware/vmware_guest_snapshot.py

##### ANSIBLE VERSION
```
2.4devel
```

### CODE REVIEW

**Comment by Akasurde** [comment]
After fix : 

```
changed: [localhost] => {
    "changed": true, 
    "failed": false, 
    "invocation": {
        "module_args": {
            "datacenter": "ha-datacenter", 
            "description": "", 
            "folder": "/vm", 
            "hostname": "192.168.122.68", 
            "memory": true, 
            "name": "SampleVM", 
            "name_match": "first", 
            "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", 
            "quiesce": true, 
            "snapshot_name": "SampleVM_snap_0001", 
            "state": "present", 
            "username": "root", 
            "uuid": null, 
            "validate_certs": false
        }
    }
}
```

Snapshot created successfully.

File: lib/ansible/modules/cloud/vmware/vmware_guest_snapshot.py [hunk 1]
```diff
@@ -209,6 +240,13 @@ def get_snapshots_by_name_recursively(self, snapshots, snapname):
     def snapshot_vm(self, vm):
         dump_memory = False
         quiesce = False
+        # Check if Virtual Machine provides capabilities for Quiesce and Memory
+        # Snapshots
+        if vm.capability.quiescedSnapshotsSupported:
+            quiesce = self.module.params['quiesce']
+        if vm.capability.memorySnapshotsSupported:
+            dump_memory = self.module.params['memory']
```
Comment (nerzhul) [inline lib/ansible/modules/cloud/vmware/vmware_guest_snapshot.py]: i prefer if you call it memory_dump it's more understandable

Comment (Akasurde) [reply lib/ansible/modules/cloud/vmware/vmware_guest_snapshot.py]: @nerzhul Done.

**Comment by nerzhul** [comment]
ship_it

File: lib/ansible/modules/cloud/vmware/vmware_guest_snapshot.py [hunk 2]
```diff
@@ -283,6 +321,8 @@ def main():
             datacenter=dict(required=True, type='str'),
             snapshot_name=dict(required=False, type='str'),
             description=dict(required=False, type='str', default=''),
+            quiesce=dict(type='bool', default=False),
+            memory=dict(type='bool', default=False),
```
Comment (nerzhul) [inline lib/ansible/modules/cloud/vmware/vmware_guest_snapshot.py]: oh i mean the option for end users, mainly :)

**Review by nerzhul — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/cloud/vmware/vmware_guest_snapshot.py (hunk 2 above)
Comment (Akasurde) [reply lib/ansible/modules/cloud/vmware/vmware_guest_snapshot.py]: @nerzhul Done.

**Comment by nerzhul** [comment]
ship_it

**Comment by Akasurde** [comment]
@jctanner @nerzhul Thanks for review and comments.

**Pull request merged**